### ISSUE TITLE
Macro for disabling/debugging internal checks?

### ISSUE THREAD

**Issue by HechtiDerLachs** [oscar-system/Oscar.jl#1954]
In some parts of Oscar, we are using optional `check` arguments in functions like
```julia
function my_fun(a::Any, b::Any; check::Bool=true)
...
if check
  # do some expensive sanity checks on the input
end
# proceed with the actual task 
...
end
```
The idea is to check all input for validity and sanity by default. The knowledgeable user has the possibility to disable the expensive checks to speed up their computation. Similarly, for internal use, we can expect the input data to be sane and just call the functions like `my_fun` with `check=false`. 

Now I have found this very difficult to handle. When I implement stuff, I almost always forget to call functions internally with `check=false` and then big chunks of code suddenly run very slow because of this. 

But these forgotten `check=false` are quite hard to catch. So I was thinking: Would it make sense to introduce a macro like `@internal_checks` or something which also evaluates certain global flags to steer the behavior of all internal checks? Maybe this could then be used like
```julia
if @internal_checks
  # Do the expensive checks
end
```
What I would like to be able to do: When debugging, set some global flag like `WARN_INTERNAL_CHECKS=true` and then every call to `@internal_checks` prints a warning with a stacktrace if a function is called with `check=true`. 

Another possible flag would be `DISABLE_INTERNAL_CHECKS` to turn them off globally. 

Unfortunately, I have no idea yet, how to design and write macros. So my questions are: 

 1. Do you think this is feasible?
 2. Would you also like to make use of this? Resp.: Do you think, it makes sense?

**Comment by thofma** [comment]
> Now I have found this very difficult to handle. When I implement stuff, I almost always forget to call functions internally > with `check=false` and then big chunks of code suddenly run very slow because of this. 

What do you mean with "suddenly"? So there was a version with `check = false` and you removed it and it gotten suddenly slower?

I understand the problem, but I can't follow the logic of the `@internal_checks`. A callee cannot know whether a check is internal or not. Only the caller has this information. The concept of "internal check" is not something intrinsic to a function.

Why not have `@check`, which turns every
```
if @check
  expensive_check()
end
```
into
```
if check || WARN_INTERNAL_CHECK
  @warn "Check called in ..."
  expensive_check()
```
One can make this a bit more clever/efficient, but in principle this has no overhead. It should do what you want if `WARN_INTERNAL_CHECK == true`?

**Comment by fieker** [comment]
If we allow a global check=false, we may stop implementing checks directly.
We can discuss this, but a global check=false is going to produce an
interesting collection of bug reports, stemming from wrong input
However for debugging purposes, consider @hassert....

**Comment by thofma** [comment]
I don't think it is about a global `check = false`. It is mainly about the following situation. Assume I write a function in Oscar:
```julia
function do_something_amazing(a, b)
  check_wether_input_makes_sense(a, b)
  K = number_field(a)
  I = some_complicated_ideal_sheaf(b)
end
```
Since I checked that `a` and `b` are valid input, I know that they are valid input for the rest. In particular, I can do `check = false` in the subsequent functions, e.g. I can do `number_field(a, check = false)` and similar for the other call. The problem is, how do I know for which functions I can actually provide `check = false`? This is impossible to find out at the moment. So the suggestion is to set  `WARN_INTERNAL_CHECK = true` (just for the purpose finding these things), which will tell me
```
Warning: number_field() called with `check = true` at blub.jl:1234.
Warning: some_complicated_ideal_sheaf called with `check = true` at blub.jl:1235.
```
I can now go and check whether it makes sense to turn that of.

@HechtiDerLachs I hope I explained the intention of your proposal correctly.

**Comment by simonbrandhorst** [comment]
Looks a bit like `@inbounds` to me.
